### ISSUE TITLE
Postpone bulky header data until when it's needed

### ISSUE THREAD

**Issue by jonsneyers** [FLIF-hub/FLIF#119]
In particular, the MANIAC trees and the ACB color data is now outputted all at once in the header, while it could in principle also be outputted "on-demand" (i.e. tree nodes when their count reaches zero, ACB buckets when they are first used).

Disadvantage: memory for header data coders has to remain allocated until (almost) the end. As it is now, it can be freed once the header is encoded/decoded. On the other hand, memory for the MANIAC trees will be saved, especially during a partial decode.

Advantage: while the total filesize would remain the same, progressive decoding would be improved. Especially for large images, which may very well have 100kb or more of header data, the difference would be big.

This would require some code changes and it would be a big change to the decode stream, so either this has to be done now (before M1), or it has to be postponed and both methods (the current "big header" and the new "fragmented header") have to remain supported, with the new method only being available in newer versions.
I'm leaning towards doing it now.

**Comment by sipa** [comment]
ACK on doing that now. It will improve the quality/size ratio for very
strong zoom levels, I expect.

**Comment by jonsneyers** [comment]
OK, I've given this some more thought.

"Lazy loading" of the MANIAC trees means that they have to be encoded/decoded in a kind of breadth-first way instead of depth-first. That means there is a bit of a decode speed penalty to be paid, because now the property ranges (which get narrowed while traversing the tree) only need to be computed/maintained once during the tree loading in the decoder, while if we would scatter the tree all over the file in a "lazy loading" manner, then those property ranges have to be recomputed by traversing the tree again each time a split happens.

Also it would complicate the code, since now we simply have one implementation of find_leaf() in FinalPropertySymbolCoder, which can be used both during (the final step of) encoding and decoding, while if we change the tree loading from eager (header) to lazy (scattered all over the file), we would need a different find_leaf() for decoding and encoding.

So I'm not really convinced that this is a good thing to do. I would rather try to tweak the tree encoding a bit in order to keep the header overhead acceptable.

The ACB/PLT/PLA header overhead (and for animations you can add FRS and FRA) is more important anyway, since that information appears in the stream even before the "rough preview" zoomlevels. Of course those transformations are optional, so if rapid previews are important, they can be disabled, but still...
Theoretically the ACB/PLT/PLA data could be transmitted in a lazy way too: only output a palette color or ACB bucket at the time it is actually needed. However, that would probably come at the price of a significant slowdown, since it would introduce extra branching in the inner loops. Also it would break the modular design of the transforms.

So, in conclusion, I think this is too much implementation effort and (probably) speed sacrifice for too little benefit. For extremely large images, some kind of hierarchical tiles (cf #12) are going to be needed anyway (otherwise cropped zooming takes too long), and for moderate sized images, the header overhead is reasonable enough. Also, if very quick previewing is really important, you can encode using something like `flif -bp0 -S200` and get a small header.
